src/lib.rs

```diff
@@ -323,8 +326,45 @@ impl fmt::Display for GlobError {
     }
 }
 
-fn is_dir(p: &Path) -> bool {
-    fs::metadata(p).map(|m| m.is_dir()).unwrap_or(false)
+#[derive(Debug)]
+struct PathWrapper {
+    path: PathBuf,
+    is_dir: Option<bool>,
+}
+
+impl PathWrapper {
+    fn from_dir_entry(path: PathBuf, e: DirEntry) -> Self {
+        Self {
+            path,
+            is_dir: e.file_type().ok().map(|file_type| file_type.is_dir()),
+        }
+    }
+    fn from_path(path: PathBuf) -> Self {
+        Self { path, is_dir: None }
+    }
+
+    fn into_path(self) -> PathBuf {
+        self.path
+    }
+}
+
+impl Deref for PathWrapper {
+    type Target = Path;
+
+    fn deref(&self) -> &Self::Target {
+        self.path.deref()
+    }
+}
+
+impl AsRef<Path> for PathWrapper {
+    fn as_ref(&self) -> &Path {
+        self.path.as_ref()
+    }
+}
+
+fn is_dir(p: &PathWrapper) -> bool {
+    p.is_dir
+        .unwrap_or_else(|| fs::metadata(&p.path).map(|m| m.is_dir()).unwrap_or(false))
```

Comment:
The cache will be missed if `metadata` has to be called, yeah. I guess that we could actually call `metadata` eagerly when the path is constructed (if `DirEntry` isn't available), since eventually the code will probably call `is_dir` on each found path.

Reply:
I changed it so that the information about being a directory is always computed. It is taken from `DirEntry` if possible, and if not, `fs::metadata` is called, but always at most once for a given path.